contracts/token/ERC1155/ERC1155MetadataURICatchAll.sol

```diff
@@ -0,0 +1,46 @@
+pragma solidity ^0.6.0;
+
+import "./ERC1155.sol";
+import "./IERC1155MetadataURI.sol";
+import "../../introspection/ERC165.sol";
+
+contract ERC1155MetadataURICatchAll is ERC165, ERC1155, IERC1155MetadataURI {
+    // Catch-all URI with placeholders, e.g. https://example.com/{locale}/{id}.json
+    string private _uri;
+
+     /*
+     *     bytes4(keccak256('uri(uint256)')) == 0x0e89341c
+     */
+    bytes4 private constant _INTERFACE_ID_ERC1155_METADATA_URI = 0x0e89341c;
+
+    /**
+     * @dev Constructor function
+     */
+    constructor (string memory uri) public {
+        _setURI(uri);
+
+        // register the supported interfaces to conform to ERC1155 via ERC165
+        _registerInterface(_INTERFACE_ID_ERC1155_METADATA_URI);
+    }
+
+    /**
+     * @notice A distinct Uniform Resource Identifier (URI) for a given token.
+     * @dev URIs are defined in RFC 3986.
+     * The URI MUST point to a JSON file that conforms to the "ERC-1155 Metadata URI JSON Schema".
+     * param id uint256 ID of the token to query (ignored in this particular implementation,
+     * as an {id} parameter in the string is expected)
+     * @return URI string
+    */
+    function uri(uint256 /*id*/) external view override returns (string memory) {
+        return _uri;
+    }
+
+    /**
+     * @dev Internal function to set a new URI
+     * @param newuri New URI to be set
+     */
+    function _setURI(string memory newuri) internal virtual {
+        _uri = newuri;
```

Comment:
What do you think about a runetime check validating the existance of the `{id}` substring? @frangio this is an interesting use case for string utils.

Reply:
Well, you could in theory set a completely static URI that points to the same JSON document for all IDs. There is nothing in the metadata that specifically has to be different for different IDs or unique to one ID. In practice, users will probably always want the ID in the URI, but it's not completely required.
If there is a string util for detecting if a string is contained in another, I'd be fine with adding such a check but I don't want to code that test myself. ;-)

Reply:
True, though I'm not sure a constant URI for all ids is very useful :stuck_out_tongue: I was just wondering if people would consider this helpful or annoying.

Reply:
It's clearly optional in the EIP so I wouldn't enforce it.
>If the string {id} exists in any URI, clients MUST replace this with the actual token ID in hexadecimal form.